src/property_estimator_io.py

```diff
@@ -0,0 +1,731 @@
+
+        Returns
+        -------
+        np.ndarray
+            The array of values of shape (len(self._gradient_key_mappings),)
+        """
+
+        parameter_values = np.zeros(len(self._gradient_key_mappings))
+
+        for gradient_key, parameter_index in self._gradient_key_mappings.items():
+
+            parameter_value, _ = self._parameter_value_from_gradient_key(gradient_key)
+            expected_unit = self._parameter_units[gradient_key]
+
+            parameter_values[parameter_index] = parameter_value.to(expected_unit).magnitude
+
+        return parameter_values
+
+    def _build_pvals_jacobian(self, mvals, perturbation_amount=1.0e-4):
+        """Build the matrix which maps the gradients of properties with
+        respect to physical parameters to gradients with respect to
+        force balance mathematical parameters.
+
+        Parameters
+        ----------
+        mvals: np.ndarray
+            The current force balance mathematical parameters.
+        perturbation_amount: float
+            The amount to perturb the mathematical parameters by
+            when calculating the finite difference gradients.
+
+        Returns
+        -------
+        np.ndarray
+            A matrix of d(Physical Parameter)/d(Mathematical Parameter).
+        """
+
+        jacobian_list = []
+
+        for index in range(len(mvals)):
+
+            reverse_mvals = mvals.copy()
+            reverse_mvals[index] -= perturbation_amount
+
+            self.FF.make(reverse_mvals)
+            reverse_physical_values = self._extract_physical_parameter_values()
+
+            forward_mvals = mvals.copy()
+            forward_mvals[index] += perturbation_amount
+
+            self.FF.make(forward_mvals)
+            forward_physical_values = self._extract_physical_parameter_values()
+
+            gradients = (forward_physical_values - reverse_physical_values) / (2.0 * perturbation_amount)
+            jacobian_list.append(gradients)
+
+        # Make sure to restore the FF object back to its original state.
+        self.FF.make(mvals)
+
+        jacobian = np.array(jacobian_list)
+        return jacobian
+
+    def submit_jobs(self, mvals, AGrad=True, AHess=True):
+        """
+        Submit jobs for evaluating the objective function
+
+        Parameters
+        ----------
+        mvals: np.ndarray
+            mvals array containing the math values of the parameters
+        AGrad: bool
+            Flag for computing gradients of not
+        AHess: bool
+            Flag for computing hessian or not
+
+        Notes
+        -----
+        1. This function is called before wq_complete() and get().
+        2. This function should not block.
+        """
+
+        # Make the force field based on the current values of the parameters.
+        self.FF.make(mvals)
+
+        force_field = smirnoff.ForceField(self.FF.offxml, allow_cosmetic_attributes=True)
+
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.offxml') as file:
```

Comment:
(not blocking, could become issue) It may be worthwhile to replace the `NamedTemporaryFile` here with a `StringIO` object, so that a slow file system won't have a negative effect here.

Reply:
I agree. `SpooledTemporaryFile` probably also works here.

Reply:
Actually, does `force_field.to_file()` and `smirnoff.ForceField()` support file-like objects in addition to a file path?

Reply:
I think `ForceField.to_string()` could work well here. Here's the relevant test: https://github.com/openforcefield/openforcefield/blob/master/openforcefield/tests/test_forcefield.py#L419-L452